// streetjs/dist/release/scorecard.js
// Release Engineering — scorecard, validation, and health-delta logic (zero runtime deps).
//
// This module holds the pure decision logic for the Release Engineering
// subsystem (Requirement 11). It lives in `@streetjs/core` so the same helpers
// can be reused by the CLI/scripts renderer and exercised by property tests
// without third-party dependencies — it uses only language built-ins.
//
//   - `isValidSemver()`        validate MAJOR.MINOR.PATCH (Req 11.2)
//   - `validateReleaseNotes()` confirm a non-empty changelog entry (Req 11.2)
//   - `ReleaseScorecard`       bounded 0–100 scores (Req 11.1)
//   - `HealthMetrics`          counts + deltas (current − previous) (Req 11.4)
//   - `buildReleaseReport()`   assemble the report, recording the failed
//                              control when a validation fails (Req 11.3/11.5)
//
// _Requirements: 11.1, 11.2, 11.4, 11.5_
/** The lowest permitted score on the scorecard scale. */
export const MIN_SCORE = 0;
/** The highest permitted score on the scorecard scale. */
export const MAX_SCORE = 100;
/**
 * Strict semver MAJOR.MINOR.PATCH core: three numeric identifiers with no
 * leading zeros and no prerelease/build metadata (Requirement 11.2).
 */
const SEMVER_CORE_PATTERN = /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)$/;
/**
 * Returns true iff `version` is a valid semver MAJOR.MINOR.PATCH string —
 * exactly three dot-separated numeric identifiers, each without leading zeros
 * (Requirement 11.2). Pure and deterministic.
 */
export function isValidSemver(version) {
    if (typeof version !== 'string')
        return false;
    return SEMVER_CORE_PATTERN.test(version);
}
/** Escape a string for safe use inside a `RegExp`. */
function escapeRegExp(value) {
    return value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}
/**
 * Returns true iff `changelog` contains a non-empty notes entry for `version`
 * (Requirement 11.2). An entry is a markdown heading line (`#`/`##`/…) that
 * mentions the version, followed by at least one line of non-blank,
 * non-heading content before the next heading or end of file.
 *
 * The version match is delimited so `1.2.3` does not match `1.2.30`. Pure and
 * deterministic.
 */
export function validateReleaseNotes(changelog, version) {
    if (typeof changelog !== 'string' || changelog.length === 0)
        return false;
    if (!isValidSemver(version))
        return false;
    const lines = changelog.split(/\r?\n/);
    // Heading lines that reference the version, delimited so 1.2.3 ≠ 1.2.30.
    const versionRef = new RegExp(`(?<![\\d.])${escapeRegExp(version)}(?![\\d.])`);
    const headingPattern = /^\s{0,3}#{1,6}\s/;
    for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        if (!headingPattern.test(line) || !versionRef.test(line))
            continue;
        // Found the version's heading; scan its body until the next heading.
        for (let j = i + 1; j < lines.length; j++) {
            const body = lines[j];
            if (headingPattern.test(body))
                break; // next section begins
            if (body.trim().length > 0)
                return true; // non-empty content found
        }
    }
    return false;
}
/**
 * Clamp a single score to `[MIN_SCORE, MAX_SCORE]`. Non-finite values (NaN,
 * ±Infinity) collapse to `MIN_SCORE` so the result is always a valid bounded
 * number.
 */
export function clampScore(value) {
    if (typeof value !== 'number' || Number.isNaN(value))
        return MIN_SCORE;
    if (value < MIN_SCORE)
        return MIN_SCORE;
    if (value > MAX_SCORE)
        return MAX_SCORE;
    return value;
}
/**
 * Return a copy of `scorecard` with every dimension clamped to `[0, 100]`
 * (Requirement 11.1). Guarantees the produced scores are bounded regardless of
 * the raw inputs. Pure and deterministic.
 */
export function boundScorecard(scorecard) {
    return {
        security: clampScore(scorecard.security),
        reliability: clampScore(scorecard.reliability),
        coverage: clampScore(scorecard.coverage),
        performance: clampScore(scorecard.performance),
    };
}
/** Compute a {@link HealthMetric} from current and previous counts (`current − previous`). */
function healthDelta(current, previous) {
    return { count: current, deltaVsPrevious: current - previous };
}
/**
 * Compute the {@link HealthMetrics} block from current and previous counts. Each
 * delta is exactly `current − previous` (Requirement 11.4). Pure and deterministic.
 */
export function computeHealthMetrics(current, previous) {
    return {
        dependencyFreshness: healthDelta(current.dependencyFreshness, previous.dependencyFreshness),
        testTrends: healthDelta(current.testTrends, previous.testTrends),
        vulnerabilityTrends: healthDelta(current.vulnerabilityTrends, previous.vulnerabilityTrends),
    };
}
/**
 * Build the release report (Requirement 11.5): a bounded scorecard, the semver
 * and release-notes validation outcomes, and the health deltas. When a control
 * fails, the report records the failed control (`'semver'` first, then
 * `'release-notes'`) so CI can fail the release indicating which validation
 * failed (Requirement 11.3). Pure and deterministic given an explicit timestamp.
 */
export function buildReleaseReport(input) {
    const semverOk = isValidSemver(input.version);
    const releaseNotesOk = validateReleaseNotes(input.changelog, input.version);
    const validation = { semverOk, releaseNotesOk };
    if (!semverOk) {
        validation.failedControl = 'semver';
    }
    else if (!releaseNotesOk) {
        validation.failedControl = 'release-notes';
    }
    return {
        version: input.version,
        scorecard: boundScorecard(input.scorecard),
        validation,
        health: computeHealthMetrics(input.health.current, input.health.previous),
        timestamp: input.timestamp ?? new Date().toISOString(),
    };
}
