// streetjs/dist/security/abuse.js
// src/security/abuse.ts
// Phase 6 — Abuse_Engine (Requirement 7).
//
// A counter-backed abuse-prevention engine consulted by the authentication path
// (composes the JWT/session/auth primitives). It is built entirely over the
// pluggable `CounterStore` sliding-window abstraction from `store.ts`, so it
// runs in-memory by default and can be backed by a shared external store for
// consistent enforcement across multiple application instances.
//
// Responsibilities (mapped to acceptance criteria):
//   - Account lockout after repeated failed logins within a window (R7.1/R7.2).
//   - Per-source signup throttling within a window (R7.3).
//   - Password-spray classification across distinct accounts (R7.4).
//   - A composite suspicious-activity score from configured signals (R7.5),
//     including the IP-reputation hook (R7.7), with a configured response
//     action triggered once the score threshold is reached (R7.6).
//
// All time inputs are explicit milliseconds; an injected clock supplies the
// default "now" so window timing is fully deterministic under test.
import { systemClock } from './store.js';
/**
 * Counter-backed abuse-prevention engine (R7).
 *
 * The engine derives every stateful decision from sliding-window counters in
 * the injected {@link CounterStore}; no per-account or per-IP mutable state is
 * held on the instance itself, so the same store can be shared across
 * application instances for consistent enforcement.
 */
export class AbuseEngine {
    cfg;
    store;
    ipReputation;
    clock;
    constructor(cfg, store, ipReputation, opts = {}) {
        this.cfg = cfg;
        this.store = store;
        this.ipReputation = ipReputation;
        this.clock = opts.clock ?? systemClock;
    }
    // --- Key helpers (namespaced so a shared store can host many subsystems) ---
    loginFailKey(accountId) {
        return `abuse:login-fail:${accountId}`;
    }
    lockoutKey(accountId) {
        return `abuse:lockout:${accountId}`;
    }
    signupKey(ip) {
        return `abuse:signup:${ip}`;
    }
    /** Per-(source, account) marker used to count an account only once per window. */
    sprayPairKey(ip, accountId) {
        return `abuse:spray-pair:${ip}:${accountId}`;
    }
    /** Per-source distinct-account counter feeding spray classification. */
    sprayDistinctKey(ip) {
        return `abuse:spray-distinct:${ip}`;
    }
    /**
     * Record a login attempt and return a structured decision (R7.1/R7.2/R7.5/R7.6).
     *
     * Order of evaluation:
     *  1. If the account is already locked out, refuse immediately (R7.2).
     *  2. On a failed attempt, record the failure, advance spray tracking, and
     *     trip a lockout once the failure threshold is reached (R7.1/R7.4).
     *  3. If the attempt just tripped (or is under) a lockout, refuse (R7.2).
     *  4. Otherwise compute the score; if it reaches the threshold, trigger the
     *     configured response action and refuse (R7.5/R7.6).
     */
    async recordLoginAttempt(signal) {
        const { ip, accountId, failed, ts } = signal;
        if (accountId && (await this.isLockedOut(accountId, ts))) {
            return {
                allowed: false,
                reason: 'LOCKED_OUT',
                retryAfterMs: this.cfg.lockoutMs,
                score: await this.score(signal),
            };
        }
        if (failed && accountId) {
            const failCount = await this.store.increment(this.loginFailKey(accountId), ts, this.cfg.loginWindowMs);
            // Count this account toward the source's distinct-account spray total only
            // the first time it is seen within the spray window.
            const pairCount = await this.store.increment(this.sprayPairKey(ip, accountId), ts, this.cfg.sprayWindowMs);
            if (pairCount === 1) {
                await this.store.increment(this.sprayDistinctKey(ip), ts, this.cfg.sprayWindowMs);
            }
            // Trip the lockout once the failure threshold is reached (R7.1).
            if (failCount >= this.cfg.loginFailureThreshold) {
                await this.store.increment(this.lockoutKey(accountId), ts, this.cfg.lockoutMs);
            }
        }
        // The failure above may have just tripped a lockout; refuse if so (R7.2).
        if (accountId && (await this.isLockedOut(accountId, ts))) {
            return {
                allowed: false,
                reason: 'LOCKED_OUT',
                retryAfterMs: this.cfg.lockoutMs,
                score: await this.score(signal),
            };
        }
        const score = await this.score(signal);
        if (score >= this.cfg.scoreThreshold) {
            const decision = { allowed: false, reason: 'SCORE_EXCEEDED', score };
            await this.cfg.responseAction?.(decision);
            return decision;
        }
        return { allowed: true, score };
    }
    /**
     * Record a signup attempt from a source and return a decision (R7.3).
     *
     * Once the per-source attempt count reaches the configured threshold within
     * the window, further attempts from that source are throttled.
     */
    async recordSignupAttempt(ip, ts) {
        const count = await this.store.increment(this.signupKey(ip), ts, this.cfg.signupWindowMs);
        if (count >= this.cfg.signupThreshold) {
            return {
                allowed: false,
                reason: 'SIGNUP_THROTTLED',
                retryAfterMs: this.cfg.signupWindowMs,
                score: 0,
            };
        }
        return { allowed: true, score: 0 };
    }
    /**
     * Whether `accountId` is currently in Account_Lockout (R7.2). A lockout marker
     * is recorded with a {@link AbuseConfig.lockoutMs} window, so the account is
     * considered locked out while any marker remains within that window.
     */
    async isLockedOut(accountId, now = this.clock()) {
        const active = await this.store.count(this.lockoutKey(accountId), now, this.cfg.lockoutMs);
        return active > 0;
    }
    /**
     * Whether failed logins from `ip` span at least the configured number of
     * distinct accounts within the spray window, classifying the activity as a
     * password-spray pattern (R7.4).
     */
    async detectPasswordSpray(ip, now = this.clock()) {
        const distinct = await this.store.count(this.sprayDistinctKey(ip), now, this.cfg.sprayWindowMs);
        return distinct >= this.cfg.sprayDistinctAccounts;
    }
    /**
     * Compute a suspicious-activity score from configured signals (R7.5),
     * consulting the IP-reputation hook when configured (R7.7). The score is a
     * non-negative sum of:
     *   - recent failed-login count for the targeted account,
     *   - distinct accounts targeted from the source (spray pressure),
     *   - the IP-reputation hook's contribution.
     */
    async score(signal) {
        const { ip, accountId, ts } = signal;
        let total = 0;
        if (accountId) {
            total += await this.store.count(this.loginFailKey(accountId), ts, this.cfg.loginWindowMs);
        }
        total += await this.store.count(this.sprayDistinctKey(ip), ts, this.cfg.sprayWindowMs);
        if (this.ipReputation) {
            total += await this.ipReputation(ip);
        }
        return total;
    }
}
