// streetjs/dist/security/moderation.js
// src/security/moderation.ts
// Phase 7 — Moderation_Toolkit (Requirement 8).
//
// Report/block/mute APIs over a pluggable store, an exposed moderation queue,
// and an append-only audit log. The audit log composes the append-only pattern
// used by `auth/audit-writer.ts`: every report/block/mute/resolve action appends
// an immutable Audit_Event, and the public API exposes only append + list — there
// is no mutation path through which a recorded event can be altered (R8.5/R8.7).
//
// Responsibilities (mapped to acceptance criteria):
//   - Submitting a report stores it and places it in the moderation queue (R8.1).
//   - Blocking records the block relationship (R8.2).
//   - While A has blocked B, B cannot message A (R8.3).
//   - Muting suppresses a muted user's content for the muting user only, while
//     preserving it for other recipients (R8.4).
//   - Every report/block/mute/resolve appends an Audit_Event recording the
//     actor, target, action, and timestamp (R8.5).
//   - The moderation queue can be listed and each report resolved (R8.6).
//   - Audit_Events are not modifiable through the public moderation API (R8.7).
//
// All time inputs flow through an injected clock so timestamps are deterministic
// under test, mirroring the sibling `abuse.ts` / `store.ts` modules.
import { randomUUID } from 'node:crypto';
import { systemClock } from './store.js';
/** Deep-freeze an audit event so neither callers nor the store can mutate it. */
function freezeEvent(e) {
    return Object.freeze({ ...e });
}
/**
 * Default in-memory {@link ModerationStore}.
 *
 * Audit events are stored frozen and `listAudit` returns a fresh array of those
 * frozen events, so neither mutating the returned array nor its elements affects
 * stored state — the only way to add an event is `appendAudit` (R8.7).
 */
export class InMemoryModerationStore {
    audit = [];
    reports = new Map();
    // Set of `${blocker}\u0000${blocked}` keys.
    blocks = new Set();
    // Set of `${muter}\u0000${muted}` keys.
    mutes = new Set();
    static pairKey(a, b) {
        return `${a}\u0000${b}`;
    }
    async appendAudit(e) {
        this.audit.push(freezeEvent(e));
    }
    async listAudit() {
        // Return a copy of the (frozen) events so the internal log is not exposed.
        return this.audit.slice();
    }
    async saveReport(r) {
        this.reports.set(r.id, { ...r });
    }
    async getReport(id) {
        const r = this.reports.get(id);
        return r ? { ...r } : undefined;
    }
    async listQueue() {
        const pending = [];
        for (const r of this.reports.values()) {
            if (!r.resolution)
                pending.push({ ...r });
        }
        return pending;
    }
    async setBlock(blocker, blocked) {
        this.blocks.add(InMemoryModerationStore.pairKey(blocker, blocked));
    }
    async isBlocked(blocker, blocked) {
        return this.blocks.has(InMemoryModerationStore.pairKey(blocker, blocked));
    }
    async setMute(muter, muted) {
        this.mutes.add(InMemoryModerationStore.pairKey(muter, muted));
    }
    async isMuted(muter, muted) {
        return this.mutes.has(InMemoryModerationStore.pairKey(muter, muted));
    }
}
/** Error thrown when an operation references a report id that does not exist. */
export class UnknownReportError extends Error {
    reportId;
    constructor(reportId) {
        super(`Unknown report: ${reportId}`);
        this.name = 'UnknownReportError';
        this.reportId = reportId;
    }
}
/**
 * The Moderation_Toolkit (R8).
 *
 * A thin, auditable layer over a {@link ModerationStore}. Every state-changing
 * operation appends an {@link AuditEvent} before returning, so the audit log is
 * a complete, append-only record of moderation activity (R8.5/R8.7).
 */
export class ModerationToolkit {
    store;
    clock;
    constructor(store = new InMemoryModerationStore(), opts = {}) {
        this.store = store;
        this.clock = opts.clock ?? systemClock;
    }
    async record(actor, target, action) {
        await this.store.appendAudit({
            id: randomUUID(),
            actor,
            target,
            action,
            ts: this.clock(),
        });
    }
    /**
     * Submit a report against another user. The report is stored and placed in
     * the moderation queue, and a `report` audit event is recorded (R8.1/R8.5).
     */
    async report(reporter, target, reason) {
        const report = {
            id: randomUUID(),
            reporter,
            target,
            reason,
            createdAt: this.clock(),
        };
        await this.store.saveReport(report);
        await this.record(reporter, target, 'report');
        return { ...report };
    }
    /**
     * Record that user `a` has blocked user `b` (R8.2), appending a `block` audit
     * event (R8.5).
     */
    async block(a, b) {
        await this.store.setBlock(a, b);
        await this.record(a, b, 'block');
    }
    /**
     * Whether `from` may send a message to `to`.
     *
     * A block relationship from A to B prevents B from messaging A (R8.3); hence
     * `from` may message `to` if and only if `to` has not blocked `from`.
     */
    async canMessage(from, to) {
        return !(await this.store.isBlocked(to, from));
    }
    /**
     * Record that `muter` has muted `muted` (R8.4), appending a `mute` audit
     * event (R8.5).
     */
    async mute(muter, muted) {
        await this.store.setMute(muter, muted);
        await this.record(muter, muted, 'mute');
    }
    /**
     * Filter a recipient's feed, suppressing items whose sender the recipient has
     * muted while leaving every other item intact (R8.4). Muting is scoped to the
     * muting user only — the same items delivered to a different recipient are
     * unaffected.
     */
    async deliverable(recipient, items) {
        const result = [];
        for (const item of items) {
            if (!(await this.store.isMuted(recipient, item.sender))) {
                result.push(item);
            }
        }
        return result;
    }
    /** List pending reports awaiting moderation (R8.6). */
    async queue() {
        return this.store.listQueue();
    }
    /**
     * Record a moderator's resolution for a queued report (R8.6), appending a
     * `resolve` audit event (R8.5). Throws {@link UnknownReportError} if the
     * report id is not known.
     */
    async resolve(moderator, reportId, outcome) {
        const report = await this.store.getReport(reportId);
        if (!report)
            throw new UnknownReportError(reportId);
        report.resolution = { moderator, outcome, resolvedAt: this.clock() };
        await this.store.saveReport(report);
        await this.record(moderator, report.target, 'resolve');
    }
    /**
     * The append-only audit log (R8.5). The returned events are immutable and the
     * toolkit exposes no method to update or delete them (R8.7).
     */
    async audit() {
        return this.store.listAudit();
    }
}
