// streetjs/dist/security/privacy.js
// src/security/privacy.ts
// Phase 9 — Privacy_Controls (Requirement 10).
//
// Account deletion, data export, retention enforcement, and consent tracking so
// the platform can meet privacy obligations (GDPR/CCPA-style). The subsystem is
// deliberately storage-agnostic: applications register a `PersonalDataSource`
// per data domain (profiles, messages, audit, ...) and the controls fan out
// across every registered source for export and erasure. Retention is driven by
// per-record-type policies evaluated against a pluggable record store with an
// injected clock so a cycle is fully deterministic under test.
//
// Responsibilities (mapped to acceptance criteria):
//   - Export the personal data held for a user across all sources (R10.1).
//   - Delete a user's personal data so subsequent reads return nothing (R10.2).
//   - Apply a configured retention policy that removes records once they exceed
//     their retention period, on the next enforcement cycle (R10.3/R10.4).
//   - Record consent grant/withdraw decisions with purpose + timestamp (R10.5),
//     where the latest decision per (user, purpose) wins, and refuse
//     purpose-dependent processing while consent is withdrawn (R10.6).
import { systemClock } from './store.js';
/** Default in-memory {@link RetentionRecordStore}. */
export class InMemoryRetentionStore {
    // Map: type -> (id -> record).
    records = new Map();
    /** Add (or replace) a retained record. */
    add(record) {
        let byId = this.records.get(record.type);
        if (!byId) {
            byId = new Map();
            this.records.set(record.type, byId);
        }
        byId.set(record.id, { ...record });
    }
    async list() {
        const out = [];
        for (const byId of this.records.values()) {
            for (const record of byId.values())
                out.push({ ...record });
        }
        return out;
    }
    async remove(type, id) {
        const byId = this.records.get(type);
        if (!byId)
            return;
        byId.delete(id);
        if (byId.size === 0)
            this.records.delete(type);
    }
    /** Number of records currently retained (primarily for tests/diagnostics). */
    size() {
        let total = 0;
        for (const byId of this.records.values())
            total += byId.size;
        return total;
    }
}
/**
 * Thrown by {@link PrivacyControls.requireConsent} when processing is attempted
 * for a purpose whose latest recorded consent decision is a withdrawal (R10.6).
 */
export class ConsentRequiredError extends Error {
    /** The user whose consent is required. */
    userId;
    /** The purpose for which consent has been withdrawn. */
    purpose;
    constructor(userId, purpose) {
        super(`Consent required for purpose "${purpose}" but it has been withdrawn`);
        this.name = 'ConsentRequiredError';
        this.userId = userId;
        this.purpose = purpose;
    }
}
/** Composite key for a consent decision: `${userId}\u0000${purpose}`. */
function consentKey(userId, purpose) {
    return `${userId}\u0000${purpose}`;
}
/**
 * Privacy controls implementing export, deletion, retention, and consent (R10).
 *
 * The controls hold no personal data themselves: export and deletion fan out
 * across registered {@link PersonalDataSource}s, and retention operates over a
 * {@link RetentionRecordStore}. Only consent decisions are retained in process,
 * keyed by (user, purpose) with the latest decision winning.
 */
export class PrivacyControls {
    sources = [];
    policies = new Map();
    retentionStore;
    clock;
    // Map: `${userId}\0${purpose}` -> latest decision for that pair.
    consent = new Map();
    constructor(opts = {}) {
        this.retentionStore = opts.retentionStore ?? new InMemoryRetentionStore();
        this.clock = opts.clock ?? systemClock;
        for (const policy of opts.policies ?? [])
            this.addRetentionPolicy(policy);
    }
    /** Register a personal-data source for export and deletion fan-out. */
    registerSource(source) {
        this.sources.push(source);
    }
    /** Add or replace the retention policy for a record type (R10.3). */
    addRetentionPolicy(policy) {
        this.policies.set(policy.recordType, { ...policy });
    }
    /**
     * Generate an export package containing the personal data the platform holds
     * for `userId`, gathered from every registered source (R10.1). Output is
     * namespaced by each source's {@link PersonalDataSource.name} so domains with
     * overlapping field names do not collide.
     */
    async exportData(userId) {
        const pkg = {};
        for (const source of this.sources) {
            pkg[source.name] = await source.collect(userId);
        }
        return pkg;
    }
    /**
     * Complete an account-deletion request: erase `userId`'s personal data from
     * every registered source so subsequent reads return no personal data (R10.2).
     */
    async deleteAccount(userId) {
        for (const source of this.sources) {
            await source.erase(userId);
        }
    }
    /**
     * Run a single retention enforcement cycle (R10.3/R10.4). A record is removed
     * when a policy exists for its type and its age (`now - createdAt`) exceeds
     * that policy's `maxAgeMs`; records without a policy, or not yet expired, are
     * retained. Returns the number of records removed in this cycle.
     */
    async enforceRetention(now = this.clock()) {
        const records = await this.retentionStore.list();
        let removed = 0;
        for (const record of records) {
            const policy = this.policies.get(record.type);
            if (!policy)
                continue;
            const age = now - record.createdAt;
            if (age > policy.maxAgeMs) {
                await this.retentionStore.remove(record.type, record.id);
                removed++;
            }
        }
        return { removed };
    }
    /**
     * Record a consent grant or withdrawal decision (R10.5). The latest decision
     * per (user, purpose) wins, determined by {@link ConsentDecision.ts}; on a tie
     * the most recently recorded decision wins.
     */
    setConsent(decision) {
        const key = consentKey(decision.userId, decision.purpose);
        const existing = this.consent.get(key);
        // Latest decision wins by timestamp; ties resolve to the newer call.
        if (!existing || decision.ts >= existing.ts) {
            this.consent.set(key, { ...decision });
        }
    }
    /**
     * Whether the latest recorded decision for (user, purpose) is a grant (R10.6).
     * Returns `false` when no decision has been recorded.
     */
    hasConsent(userId, purpose) {
        const decision = this.consent.get(consentKey(userId, purpose));
        return decision?.granted ?? false;
    }
    /**
     * Refuse purpose-dependent processing while consent is withdrawn (R10.6).
     * Throws {@link ConsentRequiredError} if and only if the latest recorded
     * decision for (user, purpose) is a withdrawal. When no decision has been
     * recorded the call passes (there is nothing to refuse).
     */
    requireConsent(userId, purpose) {
        const decision = this.consent.get(consentKey(userId, purpose));
        if (decision && !decision.granted) {
            throw new ConsentRequiredError(userId, purpose);
        }
    }
}
