// streetjs/dist/security/ratelimit.js
// src/security/ratelimit.ts
// Sliding-window rate limiter using BigInt nanosecond timestamps.
// Bounded per-key request log with periodic stale-entry sweeping.
import { randomBytes } from 'node:crypto';
import { StreetException } from '../http/exceptions.js';
import { InMemoryRateLimitStore, systemClock, } from './store.js';
const MAX_KEYS = 100_000; // max distinct IPs/keys tracked
const MAX_REQUESTS_PER_KEY = 1000; // max stored timestamps per key
// Regex for a valid IPv4 or IPv6 address segment
const VALID_IP_RE = /^[\w.:[\]]+$/;
export class RateLimiter {
    // Map: key -> array of nanosecond timestamps (bounded)
    store = new Map();
    opts;
    sweepTimer;
    constructor(opts) {
        const resolvedKeyFn = opts.keyFn ?? ((ctx) => defaultKeyFn(ctx, opts.trustProxy ?? false));
        this.opts = {
            message: 'Too Many Requests',
            trustProxy: false,
            ...opts,
            keyFn: resolvedKeyFn,
        };
        // Sweep stale keys every half-window
        const sweepInterval = Math.max(opts.windowMs / 2, 5_000);
        this.sweepTimer = setInterval(() => this._sweep(), sweepInterval);
        this.sweepTimer.unref();
    }
    middleware() {
        return async (ctx, next) => {
            const key = this.opts.keyFn(ctx);
            const allowed = this._check(key);
            const resetSeconds = Math.ceil(this.opts.windowMs / 1000);
            ctx.setHeader('X-RateLimit-Limit', String(this.opts.maxRequests));
            ctx.setHeader('X-RateLimit-Reset', String(resetSeconds));
            if (!allowed) {
                ctx.setHeader('Retry-After', String(resetSeconds));
                ctx.setHeader('X-RateLimit-Remaining', '0');
                throw new RateLimitException(this.opts.message);
            }
            const remaining = this._remaining(key);
            ctx.setHeader('X-RateLimit-Remaining', String(remaining));
            await next();
        };
    }
    _check(key) {
        const nowNs = process.hrtime.bigint();
        const windowNs = BigInt(this.opts.windowMs) * 1000000n;
        const cutoff = nowNs - windowNs;
        let timestamps = this.store.get(key);
        if (!timestamps) {
            if (this.store.size >= MAX_KEYS) {
                // Evict oldest key when at capacity
                const firstKey = this.store.keys().next().value;
                if (firstKey !== undefined)
                    this.store.delete(firstKey);
            }
            timestamps = [];
            this.store.set(key, timestamps);
        }
        // Remove expired entries
        let start = 0;
        while (start < timestamps.length && timestamps[start] < cutoff)
            start++;
        if (start > 0)
            timestamps.splice(0, start);
        if (timestamps.length >= this.opts.maxRequests)
            return false;
        // Enforce bounded per-key storage
        if (timestamps.length >= MAX_REQUESTS_PER_KEY)
            return false;
        timestamps.push(nowNs);
        return true;
    }
    _remaining(key) {
        const nowNs = process.hrtime.bigint();
        const windowNs = BigInt(this.opts.windowMs) * 1000000n;
        const cutoff = nowNs - windowNs;
        const timestamps = this.store.get(key) ?? [];
        const active = timestamps.filter((t) => t >= cutoff).length;
        return Math.max(0, this.opts.maxRequests - active);
    }
    _sweep() {
        const nowNs = process.hrtime.bigint();
        const windowNs = BigInt(this.opts.windowMs) * 1000000n;
        const cutoff = nowNs - windowNs;
        for (const [key, timestamps] of this.store.entries()) {
            const filtered = timestamps.filter((t) => t >= cutoff);
            if (filtered.length === 0) {
                this.store.delete(key);
            }
            else {
                this.store.set(key, filtered);
            }
        }
    }
    destroy() {
        clearInterval(this.sweepTimer);
        this.store.clear();
    }
}
function defaultKeyFn(ctx, trustProxy) {
    // When trustProxy is false (default), always use the direct socket address.
    // It is set by the OS/kernel and cannot be forged by the HTTP client.
    if (!trustProxy) {
        return ctx.req.socket.remoteAddress ?? 'unknown';
    }
    // trustProxy=true: take the RIGHTMOST IP added by the trusted proxy.
    // The leftmost entries in X-Forwarded-For are client-supplied and can be
    // forged; the rightmost entry is appended by the proxy we trust.
    const forwarded = ctx.headers['x-forwarded-for'];
    if (forwarded) {
        const parts = forwarded.split(',');
        // Walk right-to-left to find the first valid-looking IP
        for (let i = parts.length - 1; i >= 0; i--) {
            const candidate = parts[i]?.trim() ?? '';
            if (candidate && VALID_IP_RE.test(candidate)) {
                return candidate;
            }
        }
    }
    return ctx.req.socket.remoteAddress ?? 'unknown';
}
export class RateLimitException extends StreetException {
    constructor(message) {
        super(429, message);
        this.name = 'RateLimitException';
    }
}
/**
 * Method decorator that attaches rate-limit configuration metadata to a route
 * handler.  The actual enforcement is performed by the router middleware
 * pipeline which reads the `street:rateLimit` metadata key at dispatch time.
 *
 * @example
 * ```ts
 * @RateLimit({ requests: 100, window: 60_000 })
 * @Post('/login')
 * async login(ctx: StreetContext) { ... }
 * ```
 */
export function RateLimit(opts) {
    return (target, propertyKey) => {
        Reflect.defineMetadata('street:rateLimit', opts, target, propertyKey);
    };
}
/** Retrieve the RateLimitDecoratorOptions stored by @RateLimit, or undefined. */
export function getRateLimitMeta(target, propertyKey) {
    return Reflect.getMetadata('street:rateLimit', target, propertyKey);
}
// ─── Scoped rate limiting (R3.2–R3.8) ────────────────────────────────────────
//
// The class-based `RateLimiter` above remains for backward compatibility. The
// API below extends it with (a) a human-readable window parser, (b) explicit
// global / per-IP / per-user scopes, and (c) a pluggable `RateLimitStore`
// backing abstraction so counts can be enforced consistently across multiple
// application instances (the default `InMemoryRateLimitStore` keeps the original
// in-process sliding-window semantics; `RedisRateLimitStore` shares state via a
// sorted set per key). The existing sliding-window / `Retry-After` /
// `X-RateLimit-*` response behavior is preserved.
/** Multipliers from a duration unit to milliseconds. */
const WINDOW_UNIT_MS = {
    ms: 1,
    s: 1_000,
    m: 60_000,
    h: 3_600_000,
    d: 86_400_000,
};
// e.g. "1m", "30s", "2h", "500ms", "1.5h", or a bare "1000" (milliseconds).
// Whitespace is stripped via String.prototype.trim() BEFORE matching, so the
// pattern itself contains no \s* groups — this avoids the ambiguous
// whitespace-partition backtracking (polynomial ReDoS) that arises when
// optional \s* groups surround an optional token. The pattern is linear.
const WINDOW_RE = /^(\d+(?:\.\d+)?)(ms|s|m|h|d)?$/i;
/**
 * Parse a human-readable window duration into milliseconds (R3.7).
 *
 * Accepts a number (already milliseconds) or a string such as `"1m"`, `"30s"`,
 * `"2h"`, `"7d"`, or `"500ms"`. A bare numeric string (e.g. `"1000"`) is treated
 * as milliseconds. Throws on a non-positive or unparseable value.
 *
 * @example parseWindow("1m") // 60_000
 * @example parseWindow(5_000) // 5_000
 */
export function parseWindow(window) {
    if (typeof window === 'number') {
        if (!Number.isFinite(window) || window <= 0) {
            throw new Error(`Invalid rate-limit window: ${window}`);
        }
        return Math.floor(window);
    }
    // Trim first so the pattern needs no (ambiguous) \s* groups (ReDoS-safe).
    const match = WINDOW_RE.exec(window.trim());
    if (!match) {
        throw new Error(`Invalid rate-limit window: "${window}"`);
    }
    const value = Number.parseFloat(match[1]);
    const unit = (match[2] ?? 'ms').toLowerCase();
    const ms = value * WINDOW_UNIT_MS[unit];
    if (!Number.isFinite(ms) || ms <= 0) {
        throw new Error(`Invalid rate-limit window: "${window}"`);
    }
    return Math.floor(ms);
}
/** Fixed key used for the single shared bucket in `scope: 'global'`. */
const GLOBAL_RATE_KEY = '__global__';
/** Resolve the scoped store key for a request. */
function resolveScopedKey(opts, ctx) {
    const trustProxy = opts.trustProxy ?? false;
    switch (opts.scope) {
        case 'global':
            return `global:${GLOBAL_RATE_KEY}`;
        case 'ip':
            return `ip:${defaultKeyFn(ctx, trustProxy)}`;
        case 'user': {
            const userKey = opts.userKeyFn?.(ctx) ?? ctx.user?.id;
            if (userKey !== undefined && userKey !== '') {
                return `user:${userKey}`;
            }
            // Unauthenticated: bound by IP so the bucket is still keyed and bounded.
            return `user:ip:${defaultKeyFn(ctx, trustProxy)}`;
        }
        default: {
            // Exhaustiveness guard.
            const never = opts.scope;
            throw new Error(`Unknown rate-limit scope: ${String(never)}`);
        }
    }
}
/**
 * Middleware factory equivalent to `rateLimit({ requests: 100, window: "1m" })`
 * (R3.7), supporting global / per-IP / per-user scopes (R3.2) over a pluggable
 * sliding-window store (R3.8).
 *
 * Behavior preserved from the original `RateLimiter`: up to `requests` calls per
 * key per window are permitted; the call that would exceed the limit is rejected
 * with HTTP 429 (R3.3) and a `Retry-After` header in seconds (R3.4); permitted
 * responses carry `X-RateLimit-Remaining` with the leftover allowance (R3.5).
 *
 * @example
 * ```ts
 * router.use(rateLimit({ scope: 'ip', requests: 100, window: '1m' }));
 * ```
 */
export function rateLimit(opts) {
    if (!Number.isInteger(opts.requests) || opts.requests <= 0) {
        throw new Error(`rateLimit requires a positive integer "requests", got ${opts.requests}`);
    }
    const windowMs = parseWindow(opts.window);
    const store = opts.store ?? new InMemoryRateLimitStore();
    const clock = opts.clock ?? systemClock;
    const max = opts.requests;
    const message = opts.message ?? 'Too Many Requests';
    const resetSeconds = Math.max(1, Math.ceil(windowMs / 1000));
    return async (ctx, next) => {
        const key = resolveScopedKey(opts, ctx);
        const now = clock();
        ctx.setHeader('X-RateLimit-Limit', String(max));
        ctx.setHeader('X-RateLimit-Reset', String(resetSeconds));
        // Peek at the current window count without recording. Recording only on
        // acceptance preserves the original limiter's behavior (rejected requests
        // do not extend the window).
        const current = await store.count(key, now, windowMs);
        if (current >= max) {
            ctx.setHeader('Retry-After', String(resetSeconds));
            ctx.setHeader('X-RateLimit-Remaining', '0');
            throw new RateLimitException(message);
        }
        const count = await store.hit(key, now, windowMs);
        ctx.setHeader('X-RateLimit-Remaining', String(Math.max(0, max - count)));
        await next();
    };
}
/**
 * {@link RateLimitStore} backed by a Redis sorted set per key, enabling
 * consistent enforcement across multiple application instances (R3.8).
 *
 * Each request is stored as a sorted-set member scored by its millisecond
 * timestamp. The window is evaluated by trimming members older than
 * `nowMs - windowMs` with `ZREMRANGEBYSCORE` and counting the remainder with
 * `ZCARD`, mirroring the in-memory sliding window. A `PEXPIRE` bounds memory for
 * idle keys. Members are made unique with a random suffix so concurrent hits at
 * the same millisecond are not deduplicated.
 */
export class RedisRateLimitStore {
    redis;
    keyPrefix;
    constructor(redis, opts = {}) {
        this.redis = redis;
        this.keyPrefix = opts.keyPrefix ?? 'ratelimit:';
    }
    redisKey(key) {
        return `${this.keyPrefix}${key}`;
    }
    /** Trim members strictly older than the window's cutoff. */
    async trim(redisKey, cutoff) {
        // `(cutoff` makes the upper bound exclusive, keeping scores >= cutoff,
        // matching the in-memory store (`t >= nowMs - windowMs`).
        await this.redis.command(['ZREMRANGEBYSCORE', redisKey, '-inf', `(${cutoff}`]);
    }
    async cardinality(redisKey) {
        const card = await this.redis.command(['ZCARD', redisKey]);
        return typeof card === 'number' ? card : Number(card ?? 0);
    }
    async hit(key, nowMs, windowMs) {
        const redisKey = this.redisKey(key);
        const cutoff = nowMs - windowMs;
        await this.trim(redisKey, cutoff);
        const member = `${nowMs}-${randomBytes(8).toString('hex')}`;
        await this.redis.command(['ZADD', redisKey, nowMs, member]);
        // Bound memory for idle keys; the window can never outlive `windowMs`.
        await this.redis.command(['PEXPIRE', redisKey, Math.max(1, Math.ceil(windowMs))]);
        return this.cardinality(redisKey);
    }
    async count(key, nowMs, windowMs) {
        const redisKey = this.redisKey(key);
        const cutoff = nowMs - windowMs;
        await this.trim(redisKey, cutoff);
        return this.cardinality(redisKey);
    }
}
