// streetjs/dist/security/secret-provider.js
// src/security/secret-provider.ts
// Phase 8 — SecretProvider (Requirement 9).
//
// A single `SecretProvider` interface with first-class adapters for GitHub
// Secrets, AWS Secrets Manager, Azure Key Vault, and GCP Secret Manager (R9.2),
// plus a redaction registry that keeps retrieved secret values out of log output
// (R9.4) and a required-secret startup gate that fails fast emitting only the
// missing secret's NAME, never its value (R9.5).
//
// This module *builds on* the existing infrastructure rather than replacing it
// (R9.1):
//   - The AWS/Azure/GCP adapters delegate network retrieval to the already
//     hardened, SDK-free providers in `../cloud/secret-providers.js`, wrapping
//     them with refresh-on-read semantics and automatic redaction registration.
//   - The startup gate mirrors the required-variable failure behavior of
//     `vault.loadConfig` (emit the name, never the value, exit non-zero).
//
// Refresh-on-read (R9.6): by default every adapter re-reads its upstream on each
// `get()` (TTL = 0), so a value rotated in the external store is observed on the
// next request without a process restart. A positive `ttlMs` may be supplied to
// trade a short staleness window for fewer upstream calls; even then a rotated
// value appears once the (short) TTL elapses, never requiring a restart.
import { AwsSecretsManagerProvider as CloudAwsSecretsManagerProvider, AzureKeyVaultProvider as CloudAzureKeyVaultProvider, GcpSecretManagerProvider as CloudGcpSecretManagerProvider, } from '../cloud/secret-providers.js';
// ── Redaction registry (R9.4) ─────────────────────────────────────────────────
/**
 * The set of secret values that must never appear in log output. Values are
 * registered automatically whenever an adapter or {@link requireSecrets}
 * successfully retrieves a secret, and {@link redact} masks them in any string
 * before it is written to a log sink — including startup error handlers.
 */
const REDACTION_REGISTRY = new Set();
/** The token substituted in place of a registered secret value. */
export const REDACTION_PLACEHOLDER = '[REDACTED]';
/**
 * Register a secret value so it is masked by {@link redact} everywhere (R9.4).
 * Empty values are ignored (there is nothing meaningful to mask, and masking the
 * empty string would corrupt every log line).
 */
export function registerSecretForRedaction(value) {
    if (typeof value === 'string' && value.length > 0) {
        REDACTION_REGISTRY.add(value);
    }
}
/**
 * Replace every occurrence of a registered secret value in `line` with
 * {@link REDACTION_PLACEHOLDER} (R9.4). Uses literal substring replacement so no
 * value is interpreted as a regular expression. Returns `line` unchanged when no
 * registered secret is present.
 */
export function redact(line) {
    let out = line;
    for (const secret of REDACTION_REGISTRY) {
        if (secret.length === 0)
            continue;
        if (out.includes(secret)) {
            out = out.split(secret).join(REDACTION_PLACEHOLDER);
        }
    }
    return out;
}
/**
 * Clear the redaction registry. Intended for test isolation; production code has
 * no reason to forget secrets it has been asked to mask.
 */
export function clearRedactionRegistry() {
    REDACTION_REGISTRY.clear();
}
/**
 * Base class implementing refresh-on-read caching and automatic redaction
 * registration. Subclasses implement {@link fetchSecret} with their upstream
 * retrieval; every successfully retrieved value is registered for redaction
 * before it is returned (R9.4/R9.6).
 */
class RefreshOnReadProvider {
    ttlMs;
    now;
    cache = new Map();
    constructor(opts = {}) {
        this.ttlMs = opts.ttlMs ?? 0;
        this.now = opts.now ?? Date.now;
    }
    async get(name) {
        if (this.ttlMs > 0) {
            const cached = this.cache.get(name);
            if (cached && this.now() < cached.expiresAt)
                return cached.value;
        }
        const value = await this.fetchSecret(name);
        // Every retrieved value is masked from logs from this point on (R9.4).
        registerSecretForRedaction(value);
        if (this.ttlMs > 0) {
            this.cache.set(name, { value, expiresAt: this.now() + this.ttlMs });
        }
        return value;
    }
}
/**
 * Adapter for GitHub Actions Secrets. GitHub does not expose a read API for
 * repository/organization secrets; instead the Actions runner injects configured
 * secrets into the workflow environment. This adapter therefore resolves secrets
 * from the process environment (the GitHub-native surface), which is inherently
 * refresh-on-read since the environment is read on every call.
 */
export class GitHubSecretsProvider extends RefreshOnReadProvider {
    env;
    prefix;
    constructor(opts = {}) {
        super(opts);
        this.env = opts.env ?? process.env;
        this.prefix = opts.prefix ?? '';
    }
    async fetchSecret(name) {
        const key = `${this.prefix}${name}`;
        const value = this.env[key];
        if (value === undefined || value === '') {
            throw new Error(`GitHubSecretsProvider: secret "${name}" is not present in the environment`);
        }
        return value;
    }
}
/**
 * AWS Secrets Manager adapter. Delegates retrieval to the SDK-free cloud
 * provider (SigV4 over `node:https`) configured for refresh-on-read, or to an
 * injected {@link SecretFetcher} for testing (R9.2/R9.3/R9.6).
 */
export class AwsSecretsManagerProvider extends RefreshOnReadProvider {
    fetcher;
    constructor(opts) {
        super(opts);
        this.fetcher = opts.fetcher ?? buildCloudFetcher(new CloudAwsSecretsManagerProvider({
            region: opts.region,
            accessKeyId: opts.accessKeyId,
            secretAccessKey: opts.secretAccessKey,
            ...(opts.endpoint !== undefined ? { endpoint: opts.endpoint } : {}),
            ...(opts.tls !== undefined ? { tls: opts.tls } : {}),
            cacheTtlMs: 0,
        }));
    }
    fetchSecret(name) {
        return this.fetcher(name);
    }
}
/**
 * Azure Key Vault adapter. Delegates to the SDK-free cloud provider (bearer
 * token over `node:https`) configured for refresh-on-read, or to an injected
 * {@link SecretFetcher} for testing (R9.2/R9.3/R9.6).
 */
export class AzureKeyVaultProvider extends RefreshOnReadProvider {
    fetcher;
    constructor(opts) {
        super(opts);
        if (opts.fetcher) {
            this.fetcher = opts.fetcher;
        }
        else {
            if (!opts.vaultUrl) {
                throw new Error('AzureKeyVaultProvider: vaultUrl is required when no fetcher is provided');
            }
            this.fetcher = buildCloudFetcher(new CloudAzureKeyVaultProvider({
                vaultUrl: opts.vaultUrl,
                ...(opts.accessToken !== undefined ? { accessToken: opts.accessToken } : {}),
                ...(opts.tokenProvider !== undefined ? { tokenProvider: opts.tokenProvider } : {}),
                ...(opts.apiVersion !== undefined ? { apiVersion: opts.apiVersion } : {}),
                ...(opts.tls !== undefined ? { tls: opts.tls } : {}),
                cacheTtlMs: 0,
            }));
        }
    }
    fetchSecret(name) {
        return this.fetcher(name);
    }
}
/**
 * GCP Secret Manager adapter. Delegates to the SDK-free cloud provider (bearer
 * token over `node:https`) configured for refresh-on-read, or to an injected
 * {@link SecretFetcher} for testing (R9.2/R9.3/R9.6).
 */
export class GcpSecretManagerProvider extends RefreshOnReadProvider {
    fetcher;
    constructor(opts) {
        super(opts);
        if (opts.fetcher) {
            this.fetcher = opts.fetcher;
        }
        else {
            if (!opts.projectId) {
                throw new Error('GcpSecretManagerProvider: projectId is required when no fetcher is provided');
            }
            this.fetcher = buildCloudFetcher(new CloudGcpSecretManagerProvider({
                projectId: opts.projectId,
                ...(opts.serviceAccountToken !== undefined ? { serviceAccountToken: opts.serviceAccountToken } : {}),
                ...(opts.endpoint !== undefined ? { endpoint: opts.endpoint } : {}),
                ...(opts.tls !== undefined ? { tls: opts.tls } : {}),
                cacheTtlMs: 0,
            }));
        }
    }
    fetchSecret(name) {
        return this.fetcher(name);
    }
}
/** Wrap a cloud SecretProvider as a {@link SecretFetcher}. */
function buildCloudFetcher(provider) {
    return (name) => provider.get(name);
}
/**
 * Startup gate enforcing that every required secret can be retrieved (R9.5).
 *
 * Each name is fetched through `provider`. On success the value is registered
 * for redaction and returned in the result map. If any required secret cannot be
 * retrieved (missing, empty, or upstream error), the gate emits ONLY the missing
 * names — never any value or upstream error detail, which could leak a value —
 * and terminates the process with a non-zero exit code, mirroring
 * `vault.loadConfig`'s required-variable behavior.
 *
 * @returns a map of name → value for all required secrets when every one
 *   resolves. When a secret is missing the gate does not return (it exits).
 */
export async function requireSecrets(provider, names, opts = {}) {
    const resolved = {};
    const missing = [];
    for (const name of names) {
        try {
            const value = await provider.get(name);
            if (value === undefined || value === '') {
                missing.push(name);
                continue;
            }
            registerSecretForRedaction(value);
            resolved[name] = value;
        }
        catch {
            // Deliberately swallow the upstream error: it may embed a secret value.
            // Only the NAME is surfaced below (R9.5).
            missing.push(name);
        }
    }
    if (missing.length > 0) {
        const log = opts.log ?? ((message) => process.stderr.write(`${message}\n`));
        log(`Missing required secret(s): ${missing.join(', ')}`);
        const exit = opts.exit ?? ((code) => process.exit(code));
        exit(1);
    }
    return resolved;
}
