// streetjs/dist/security/store.js
// src/security/store.ts
// Pluggable backing-store abstraction (R3.8).
//
// Rate limiting, abuse counters, and similar subsystems need either in-process
// or shared cross-instance state. This module defines a small set of store
// interfaces so an in-memory implementation (the default) and a shared external
// implementation (e.g. a Redis-backed store) are interchangeable, allowing
// counts to be enforced consistently across multiple application instances.
//
// The in-memory rate-limit store extracts and reuses the Map-based
// sliding-window logic that previously lived inline in `ratelimit.ts`, so it is
// drop-in interchangeable with a future `RedisRateLimitStore`. All time inputs
// are explicit milliseconds, and an injected clock (now-provider) is supported
// so window timing is fully deterministic in tests.
/** Default clock backed by wall-clock milliseconds. */
export const systemClock = () => Date.now();
/** Default ceiling on the number of distinct keys tracked in memory. */
const DEFAULT_MAX_KEYS = 100_000;
/** Default ceiling on stored timestamps per key. */
const DEFAULT_MAX_REQUESTS_PER_KEY = 1000;
/**
 * In-memory {@link RateLimitStore} implementing a bounded sliding window over a
 * `Map<string, number[]>` of millisecond timestamps. Extracted from the
 * original `RateLimiter` so it is interchangeable with a future
 * `RedisRateLimitStore`. Window evaluation is driven entirely by the `nowMs`
 * arguments, so behavior is deterministic under an injected clock.
 */
export class InMemoryRateLimitStore {
    // Map: key -> ascending array of millisecond timestamps (bounded).
    store = new Map();
    clock;
    maxKeys;
    maxRequestsPerKey;
    retentionMs;
    sweepTimer;
    constructor(opts = {}) {
        this.clock = opts.clock ?? systemClock;
        this.maxKeys = opts.maxKeys ?? DEFAULT_MAX_KEYS;
        this.maxRequestsPerKey = opts.maxRequestsPerKey ?? DEFAULT_MAX_REQUESTS_PER_KEY;
        this.retentionMs = opts.retentionMs;
        if (opts.sweepIntervalMs && opts.sweepIntervalMs > 0 && this.retentionMs && this.retentionMs > 0) {
            this.sweepTimer = setInterval(() => this._sweep(), opts.sweepIntervalMs);
            // Do not keep the event loop alive solely for sweeping.
            this.sweepTimer.unref?.();
        }
    }
    /** Current time per the injected clock. Convenience for deterministic callers. */
    now() {
        return this.clock();
    }
    async hit(key, nowMs, windowMs) {
        const cutoff = nowMs - windowMs;
        let timestamps = this.store.get(key);
        if (!timestamps) {
            if (this.store.size >= this.maxKeys) {
                // Evict the oldest key when at capacity.
                const firstKey = this.store.keys().next().value;
                if (firstKey !== undefined)
                    this.store.delete(firstKey);
            }
            timestamps = [];
            this.store.set(key, timestamps);
        }
        this._prune(timestamps, cutoff);
        // Enforce bounded per-key storage: stop accumulating beyond the cap, but
        // still report the (capped) count so the limiter rejects appropriately.
        if (timestamps.length < this.maxRequestsPerKey) {
            timestamps.push(nowMs);
        }
        return timestamps.length;
    }
    async count(key, nowMs, windowMs) {
        const cutoff = nowMs - windowMs;
        const timestamps = this.store.get(key);
        if (!timestamps)
            return 0;
        let active = 0;
        for (const t of timestamps) {
            if (t >= cutoff)
                active++;
        }
        return active;
    }
    /** Remove all tracked state for `key`. */
    reset(key) {
        this.store.delete(key);
    }
    /** Number of keys currently tracked (primarily for tests/diagnostics). */
    size() {
        return this.store.size;
    }
    /** Stop the sweep timer (if any) and clear all tracked state. */
    destroy() {
        if (this.sweepTimer)
            clearInterval(this.sweepTimer);
        this.store.clear();
    }
    /** Drop leading timestamps older than `cutoff` (array is kept ascending). */
    _prune(timestamps, cutoff) {
        let start = 0;
        while (start < timestamps.length && timestamps[start] < cutoff)
            start++;
        if (start > 0)
            timestamps.splice(0, start);
    }
    /** Periodic memory sweep: drop timestamps older than the retention horizon. */
    _sweep() {
        if (!this.retentionMs)
            return;
        const cutoff = this.clock() - this.retentionMs;
        for (const [key, timestamps] of this.store.entries()) {
            this._prune(timestamps, cutoff);
            if (timestamps.length === 0)
                this.store.delete(key);
        }
    }
}
/**
 * In-memory {@link CounterStore} built on the same bounded sliding-window logic
 * as {@link InMemoryRateLimitStore}. Suitable as the default counter backing for
 * abuse-prevention counters until a shared external store is wired in.
 */
export class InMemoryCounterStore {
    store;
    constructor(opts = {}) {
        this.store = new InMemoryRateLimitStore(opts);
    }
    increment(key, nowMs, windowMs) {
        return this.store.hit(key, nowMs, windowMs);
    }
    count(key, nowMs, windowMs) {
        return this.store.count(key, nowMs, windowMs);
    }
    async reset(key) {
        this.store.reset(key);
    }
    /** Stop any timers and clear all tracked state. */
    destroy() {
        this.store.destroy();
    }
}
/**
 * In-memory {@link KeyValueStore} with optional per-entry TTL, evaluated lazily
 * against the injected clock on read. Default backing for small opaque values
 * until a shared external store is wired in.
 */
export class InMemoryKeyValueStore {
    store = new Map();
    clock;
    constructor(opts = {}) {
        this.clock = opts.clock ?? systemClock;
    }
    async get(key) {
        const entry = this.store.get(key);
        if (!entry)
            return undefined;
        if (entry.expiresAt !== undefined && entry.expiresAt <= this.clock()) {
            this.store.delete(key);
            return undefined;
        }
        return entry.value;
    }
    async set(key, value, ttlMs) {
        const expiresAt = ttlMs !== undefined ? this.clock() + ttlMs : undefined;
        this.store.set(key, { value, expiresAt });
    }
    async delete(key) {
        this.store.delete(key);
    }
    /** Clear all entries. */
    clear() {
        this.store.clear();
    }
}
