// streetjs/dist/observability/health.js
// src/observability/health.ts
// Health Check DSL: typed check registry, parallel execution with timeout, route registration.
// ── HealthCheckRegistry ────────────────────────────────────────────────────────
export class HealthCheckRegistry {
    _checks = new Map();
    _startTime = Date.now();
    /**
     * Register a health check function.
     * Defaults: type='liveness', timeoutMs=5000.
     */
    addCheck(name, fn, opts) {
        this._checks.set(name, {
            fn,
            type: opts?.type ?? 'liveness',
            timeoutMs: opts?.timeoutMs ?? 5000,
        });
    }
    /** Run all liveness checks in parallel. */
    runLiveness() {
        return this._run('liveness');
    }
    /** Run all readiness checks in parallel. */
    runReadiness() {
        const delayMs = parseInt(process.env['STREET_READINESS_DELAY_MS'] ?? '0', 10);
        if (delayMs > 0 && Date.now() < this._startTime + delayMs) {
            const result = {
                status: 'degraded',
                checks: {
                    readiness_delay: {
                        status: 'down',
                        durationMs: 0,
                        details: { reason: 'startup_delay', remainingMs: (this._startTime + delayMs) - Date.now() },
                    },
                },
            };
            return Promise.resolve(result);
        }
        return this._run('readiness');
    }
    // ── Private helpers ─────────────────────────────────────────────────────────
    async _run(type) {
        const entries = [...this._checks.entries()].filter(([, e]) => e.type === type);
        const results = await Promise.allSettled(entries.map(([name, entry]) => this._runOne(name, entry)));
        const checks = {};
        let anyDown = false;
        for (const result of results) {
            if (result.status === 'fulfilled') {
                const { name, checkResult, durationMs } = result.value;
                checks[name] = { ...checkResult, durationMs };
                if (checkResult.status === 'down')
                    anyDown = true;
            }
            // rejected is unlikely because _runOne catches all errors, but guard anyway
        }
        return {
            status: anyDown ? 'degraded' : 'ok',
            checks,
        };
    }
    async _runOne(name, entry) {
        const start = Date.now();
        const timeoutPromise = new Promise((resolve) => {
            const t = setTimeout(() => {
                resolve({ status: 'down', details: { reason: 'timeout' } });
            }, entry.timeoutMs);
            t.unref();
        });
        let checkResult;
        try {
            checkResult = await Promise.race([entry.fn(), timeoutPromise]);
        }
        catch (err) {
            const message = err instanceof Error ? err.message : String(err);
            checkResult = { status: 'down', details: { error: message } };
        }
        const durationMs = Date.now() - start;
        return { name, checkResult, durationMs };
    }
}
/**
 * Build a readiness check that treats the database as a *declared provisioned
 * dependency* (Requirement 2.12 / design "Lazy database initialization").
 *
 * Semantics:
 *  - no database configured/expected → `up` (immediately, without probing)
 *  - database configured + reachable → `up`
 *  - database configured + unreachable (or probe times out) → `down`
 *
 * This check is intended to be registered with `type: 'readiness'` only. Liveness
 * never registers a database check, so `/health/live` never depends on the DB.
 */
export function createDbReadinessCheck(opts) {
    const probeTimeoutMs = opts.probeTimeoutMs ?? 4000;
    return async () => {
        // No declared dependency: there is no database to be ready for.
        if (!opts.expected) {
            return { status: 'up', details: { dependency: 'postgres', state: 'not-configured' } };
        }
        // Declared dependency: gate readiness on a bounded reachability probe.
        let timer;
        const timeout = new Promise((resolve) => {
            timer = setTimeout(() => resolve({ status: 'down', details: { dependency: 'postgres', reason: 'timeout' } }), probeTimeoutMs);
            timer.unref();
        });
        const probe = opts
            .probe()
            .then(() => ({ status: 'up', details: { dependency: 'postgres', state: 'reachable' } }))
            .catch((err) => ({
            status: 'down',
            details: {
                dependency: 'postgres',
                reason: 'unreachable',
                error: err instanceof Error ? err.message : String(err),
            },
        }));
        try {
            return await Promise.race([probe, timeout]);
        }
        finally {
            if (timer)
                clearTimeout(timer);
        }
    };
}
// ── Route registration ────────────────────────────────────────────────────────
/**
 * Register GET /health/live and GET /health/ready on a StreetApp instance.
 * Responds 200 if all checks pass, 503 if any check is down.
 */
export function registerHealthRoutes(app, registry) {
    app.use(async (ctx, next) => {
        if (ctx.method === 'GET' && ctx.path === '/health/live') {
            const response = await registry.runLiveness();
            ctx.json({ ...response }, response.status === 'ok' ? 200 : 503);
            return;
        }
        if (ctx.method === 'GET' && ctx.path === '/health/ready') {
            const response = await registry.runReadiness();
            ctx.json({ ...response }, response.status === 'ok' ? 200 : 503);
            return;
        }
        await next();
    });
}
